src/stream/src/executor/sink.rs

```diff
@@ -190,79 +189,103 @@ impl<F: LogStoreFactory> SinkExecutor<F> {
+                                epoch > prev_epoch,
+                                "new epoch {} should be greater than prev epoch {}",
+                                epoch,
+                                prev_epoch
+                            );
+                            sink.begin_epoch(epoch).await?;
+                            SinkState::Writing { curr_epoch: epoch }
+                        }
+                    };
 
                     if let Err(e) = sink.write_batch(chunk.clone()).await {
                         sink.abort().await?;
                         return Err(e.into());
                     }
-                    empty_checkpoint_flag = false;
                 }
-                LogStoreReadItem::Barrier {
-                    next_epoch,
-                    is_checkpoint,
-                } => {
-                    assert!(next_epoch > epoch);
-                    if is_checkpoint {
-                        if in_transaction {
-                            if empty_checkpoint_flag {
-                                sink.abort().await?;
-                                tracing::debug!(
-                                    "transaction abort due to empty epoch, epoch: {:?}",
-                                    epoch
-                                );
-                            } else {
+                LogStoreReadItem::Barrier { is_checkpoint } => {
+                    state = match state {
+                        SinkState::Uninitialized => SinkState::Checkpointed { prev_epoch: epoch },
+                        SinkState::Writing { curr_epoch } => {
+                            assert!(
+                                epoch >= curr_epoch,
+                                "barrier epoch {} should not be below current epoch {}",
+                                epoch,
+                                curr_epoch
+                            );
+                            if is_checkpoint {
                                 let start_time = Instant::now();
                                 sink.commit().await?;
                                 sink_commit_duration_metrics
                                     .observe(start_time.elapsed().as_millis() as f64);
+                                SinkState::Checkpointed { prev_epoch: epoch }
+                            } else {
+                                SinkState::Writing { curr_epoch: epoch }
                             }
                         }
+                        SinkState::Checkpointed { prev_epoch } => {
+                            assert!(
+                                epoch > prev_epoch,
+                                "checkpoint epoch {} should be greater than prev checkpoint epoch: {}",
+                                epoch,
+                                prev_epoch
+                            );
+                            SinkState::Checkpointed { prev_epoch: epoch }
+                        }
+                    };
```

Comment:
Shall we distinguish if the barrier is a checkpoint here? A non-checkpoint barrier here will still yield a `SinkState::Checkpointed` with an advanced epoch, which is somehow confusing. Or we can rename `SinkState::Checkpointed` to sth. like `SinkState::Barriered` to make it clearer.

Reply:
For non-checkpoint barriers, we mostly ignore them because we don't have any action on them. We only do sanity check on whether the epoch is increasing monotonically. The term `checkpointed` marks that there is no data written yet in a new checkpoint. Therefore, when a barrier comes, no matter whether the barrier is a checkpoint barrier or not, if the current state is `checkpointed`, the next state is also `checkpointed`. I will add some comments to describe it.